internal/batching/batching.go

```diff
@@ -0,0 +1,263 @@
+			}
+			// Then recycle the batch, if your sure you're done with it.
+			batches.Recycle()
+		}
+	}()
+
+	// Send input to the batcher.
+	for entry := range r.Stream() { // where r is a some reader returning data.Entry
+		batcher.In() <- entry
+	}
+*/
+package batching
+
+import (
+	"context"
+	"errors"
+	"log/slog"
+	"sync"
+	"time"
+
+	"github.com/Azure/tattler/data"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var (
+	batchesPool = sync.Pool{
+		New: func() any {
+			return &Batches{}
+		},
+	}
+	batchPool = sync.Pool{
+		New: func() any {
+			return &Batch{}
+		},
+	}
+)
+
+func getBatches() Batches {
+	return *batchesPool.Get().(*Batches)
+}
+
+func getBatch() Batch {
+	return *batchPool.Get().(*Batch)
+}
+
+// Batches is a map of entry types to batches.
+type Batches map[data.SourceType]Batch
+
+// Recyle recycles the batches. It should not be used after this.
+func (b Batches) Recyle() {
+	for batchesK, batch := range b {
+		for batchK := range batch {
+			delete(batch, batchK)
+		}
+		batchPool.Put(&batch)
+		delete(b, batchesK)
+	}
+	batchPool.Put(&b)
+}
+
+// Iter returns a channel that iterates over the data. Closing ctx will stop the iteration.
+func (b Batches) Iter(ctx context.Context) <-chan data.Entry {
+	ch := make(chan data.Entry, 1)
+	go func() {
+		defer close(ch)
+		for _, batch := range b {
+			for _, d := range batch {
+				select {
+				case <-ctx.Done():
+					return
+				case ch <- d:
+				}
+			}
+		}
+	}()
+	return ch
+}
+
+// Batch is a map of UIDs to data.
+type Batch map[types.UID]data.Entry
+
+func (b *Batch) Map() map[types.UID]data.Entry {
+	if b == nil {
+		return nil
+	}
+	return *b
+}
+
+// Batcher is used to ingest data and emit batches.
+type Batcher struct {
+	timespan  time.Duration
+	current   Batches
+	batchSize int
+
+	in  <-chan data.Entry
+	out chan Batches
+
+	emitter func()
+
+	log *slog.Logger
+}
+
+// Option is a opional argument for New().
+type Option func(*Batcher) error
+
+// WithLogger sets the logger.
+func WithLogger(log *slog.Logger) Option {
+	return func(b *Batcher) error {
+		b.log = log
+		return nil
+	}
+}
+
+// WithBatchSize sets the batch size at which to emit at. So if you set this to 1000, it will
+// emit when it has 1000 items in the batch if we haven't hit the timespan. If the timespan
+// is hit, it will emit regardless of the batch size.
+func WithBatchSize(size int) Option {
+	return func(b *Batcher) error {
+		if size < 0 {
+			return errors.New("batch size must be greater than 0")
+		}
+		b.batchSize = size
+		return nil
+	}
+}
+
+// New creates a new Batcher.
+func New(ctx context.Context, in <-chan data.Entry, out chan Batches, timespan time.Duration, options ...Option) (*Batcher, error) {
+	if in == nil || out == nil {
+		return nil, errors.New("can't call Batcher.New() with a nil in or out channel")
+	}
+
+	b := &Batcher{
+		timespan: timespan,
+		in:       in,
+		out:      out,
+		log:      slog.Default(),
```

Comment:
Yep.  slog is a standard, zap isn't.  Across AKS I've found zerolog, zap, logurus, ...

Slog is made to try and allow third parties to work with it, so for zap:
https://github.com/samber/slog-zap

You could define and slog.Handler and pass it to this and it would use zap.